### ISSUE TITLE
Correctly deal with IV Subject names

### ISSUE THREAD

**Issue by robstradling** [zmap/zlint#166]
https://crt.sh/?id=204117585&opt=zlint
"ERROR: The postal code MUST NOT be included without an organization name
 ERROR: The 'Street Address' field MUST NOT be included without an organization name"

As described in issue #157, some of the "Prohibited" clauses in BR 7.1.4.2.2 are badly written, but looking at the "Optional" clauses helps to determine the intent.

Subject:postalCode is (emphasis mine) "Optional if the subject:organizationName, subject:givenName field, **or** subject:surname fields are present."  Ditto for subject:streetAddress.

**Comment by kumarde** [comment]
Ah, good catch @robstradling. I think this was implemented from an earlier version of the BRs where the text read: "Prohibited if the subject:organizationName field is absent."

The new versions, w/ givenName and surname checking are more up to date, so I'll remove any instances of these old ones in the codebase.

**Comment by kumarde** [comment]
Merged.

**Comment by robstradling** [comment]
@kumarde Thanks, that's removed the false positives.  However...

I'm curious as to why you've removed the following good lints:
  "The organization name field MUST not be included without a country name"
  "If organization is included in a subscriber certificate, either stateOrProvince or locality MUST be included"

I think those 2 lints should be reintroduced, and further lints should be added, such as...
  "The subject:givenName field MUST NOT be included without a subject:countryName field"
  "The subject:surname field MUST NOT be included without a subject:countryName field"
  "The subject:streetAddress field MUST NOT be included without a subject:organizationName, subject:givenName, or subject:surname field"
  etc, etc

**Comment by kumarde** [comment]
@robstradling for IV certs, there are a set of separate lints that cover whether the proper fields are included. These are:

`lint_cab_iv_requires_personal_name`, `lint_cert_policy_iv_requires_country`, `lint_cert_policy_iv_requires_province_or_locality`

I removed the ones you mentioned because they were, afaict, incorrect checks on _all_ subscriber certificates (as indicated by their source being 7.1.4.2.2 rather than 7.1.6.1). If you still think I should add them back in, let me know.

**Comment by robstradling** [comment]
@kumarde Ah, I see.  I hadn't spotted that you were already doing these "proper fields are included" lints when one of the CABForum-defined Policy OIDs is included in the cert.

Interestingly though, including the CABForum-defined Policy OIDs in serverAuth certs is optional for CAs, as far as the BRs and EVGs are concerned.  In practice though, most CAs are (or should be!) doing this due to Microsoft's requirement that:
`CAs must use  the following OIDs in the end-entity certificate: DV 2.23.140.1.2.1; OV 2.23.140.1.2.2; EV 2.23.140.1.1.; IV 2.23.140.1.2.3; EV Code Signing 2.23.140.1.3; Non-EV Code Signing 2.23.140.1.4.1.`
(from http://aka.ms/rootcert)

How does ZLint deal with serverAuth certs that don't contain a CABForum-defined Policy OID?
(For example, consider https://crt.sh/?id=9537&opt=zlint, which triggered some of the lints you removed in PR #168, but which now doesn't trigger any lints.  It has a subject:organizationName but no subject:countryName, subject:localityName or subject:stateOrProvinceName).

**Comment by kumarde** [comment]
@robstradling For lints that apply to those specifically with CABForum-defined policy OIDs, we first check if the OID exists, and if it doesn't, the result will be `NA`, which translates to the lint is not applicable for this certificate. Another example: for tests that apply to say all subscriber certificates, we'll return `NA` if it's a CA cert. 

Another thing we do is return `NE`, not effective, which translates to "this test was not around when this certificate was issued". 

One decision we made was that we are going to support BRs v1.4.8+, meaning anything that _used_ to apply to certificates but has been removed or updated since prior versions will likely not run. Your example certificate _would have_ triggered errors up until September 7, 2016 because of the clauses written in section 7.1.4.2.2, but since it was updated after then, ZLint only now checks these particular lints (`<subject_field>_must_appear`, `<subject_field>_must_not_appear`, etc.) for certs issued after September 7, 2016, so we don't trigger any false positives for expired certificates.

This is exactly what happened to the example certificate you pointed me to, which was issued in 2013 and expired in 2014. The tests are there in ZLint, but all returned `NE` because they were updated (with givenName and surname) after the cert had been issued.

Moving forward, what is _definitely_ in scope is to come up with a way to scope each lint by time - essentially adding a field that says "this lint existed from this time to this time, apply to certificates issued in that time", but I don't know a priori what these are and it would be a real pain to go backwards in time and read earlier versions of the BRs. 

Does this make sense? Thanks!
